File: napakalaki(ruby)/lib/player.rb
#encoding: utf-8
# To change this license header, choose License Headers in Project Properties.
# To change this template file, choose Tools | Templates
# and open the template in the editor.

module NapakalakiGame
require_relative 'treasure'
require_relative 'combat_result'
require_relative 'card_dealer'
require_relative 'dice'
require_relative 'bad_consequence'
require_relative 'monster'



class Player
  @@MAXLEVEL=10
  attr_reader :name, :dead, :canISteal, :visibleTreasures, :hiddenTreasures
  attr_accessor :pendingBadConsequence, :level, :enemy
  
  def initialize(name)
    @name=name
    @level=1
    @dead=true
    @canISteal=true
    @enemy=nil
    @hiddenTreasures=Array.new
    @visibleTreasures=Array.new
    @pendingBadConsequence=nil
  end
  
  def self.getMaxLevel
    return @@MAXLEVEL
  end
  
  def bringToLife()
    @dead=false;
  end
  
  #def isDead()
    #return @dead
  #end
  
  def getCombatLevel()
    nivel = @level
    @visibleTreasures.each do|t|
      nivel+=t.bonus 
    end
    return nivel
  end
  
  def incrementLevels(l)
    @level= @level+l
    if @level>10
      @level=Player.getMaxLevel
    end
  end
  
  def decrementLevels(l)
    @level= @level-l
    if @level<=0
      @level=1
    end
  end
  
  def setEnemy(e)
    @enemy=e
  end
  
  #def setPendingbadConsequence(b)
    #@pendingBadConsequence=b
  #end
  
  def applyPrize(m)
    nLevels=m.getLevelsGained()
    incrementLevels(nLevels)
    nTreasures=m.getTreasuresGained()
      if nTreasures>0
        @dealer=CardDealer.instance()
        for i in 0..nTreasures-1 do
                treasure = @dealer.nextTreasure()
                @hiddenTreasures.push(treasure)
        end
      end
  end
  
  def applyBadConsequence(m)
    bc = m.bc
    nLevels = bc.levels
    decrementLevels(nLevels)
    pendingBad = bc.adjustToFitTreasureLists(@visibleTreasures, @hiddenTreasures)
    @pendingBadConsequence=pendingBad
  end
  
  def canMakeTreasureVisible(t)
    aux=0
    resultado=true
    tiene_dosmanos=false
    @visibleTreasures.each do |v|
       if v.type == TreasureKind::BOTHHANDS
         tiene_dosmanos=true
       end
       if v.type== t.type
          resultado=false
       end
       if v.type==TreasureKind::ONEHAND
                aux=aux+1
       end
    end
    if t.type ==TreasureKind::ONEHAND && aux<2
       if tiene_dosmanos==true
          resultado=false
       else
          resultado=true
       end
    end
    if aux>=1 && t.type ==TreasureKind::BOTHHANDS
            resultado=false
    end
        return resultado;
  end
  
  def howManyVisibleTreasures(tKind)
    contador=0
    @visibleTreasures.each do |t|
      if (t.type == tKind)
        contador +=1
      end
    end
    return contador;
  end
  
  def dielfNoTreasures()
    if @visibleTreasures.empty? && @hiddenTreasures.empty?
      @dead=true 
    end
  end
  
  def getHiddenTreasures()
    return @hiddenTreasures
  end
  
  def getVisibleTreasures()
   return @visibleTreasures
  end
  
  def combat(m)
    myLevel = getCombatLevel
    monsterLevel = getOponentLevel(m)
        if @level> Player.getMaxLevel
            aux=CombatResult::WINGAME
        else 
          if myLevel>monsterLevel
             applyPrize(m)
              aux = CombatResult::WIN    
         else
            applyBadConsequence(m)
            if(shouldConvert)
              aux= CombatResult::LOSEANDCONVERT
            else
              aux = CombatResult::LOSE
            end
         end
        end
        return aux
  end
  
  #def setHiddenTreasures(h)
   # @hiddenTreasures=h
  #end
  
  def makeTreasureVisible(t)
    canl = canMakeTreasureVisible(t)
    if canl
      @visibleTreasures.push(t);
      @hiddenTreasures.delete(t);
    end
  end
  
  def discardVisibleTreasure(t)
    @visibleTreasures.delete(t)
    if @pendingBadConsequence!=nil && !@pendingBadConsequence.isEmpty
            @pendingBadConsequence.substractVisibleTreasure(t)
    end
        dielfNoTreasures
  end
  
  def discardHiddenTreasure(t)
    @hiddenTreasures.delete(t)
    if @pendingBadConsequence!=nil && !@pendingBadConsequence.isEmpty
            @pendingBadConsequence.substractHiddenTreasure(t)
    end
        dielfNoTreasures
  end
  
  def validState()
    ok=true
    if @pendingBadConsequence!=nil && !@pendingBadConsequence.isEmpty
      ok=false 
    end
    if @hiddenTreasures.length>4
      ok=false
    end
    return ok
  end
  
  def initTreasures()
    dealer = CardDealer.instance()
    dice=Dice.instance()
    bringToLife()
    treasure = dealer.nextTreasure()
    @hiddenTreasures.push(treasure)
    number = dice.nextNumber()
    if number>1
           treasure=dealer.nextTreasure()
           @hiddenTreasures.push(treasure)
    end
    if number==6
           treasure = dealer.nextTreasure()
           @hiddenTreasures.push(treasure)
    end
  end
  
  def stealTreasure()
    canl = canISteal()
     if canl
      canYou = @enemy.canYouGiveMeATreasure
        if canYou
          treasure = @enemy.giveMeATreasure
          @hiddenTreasures.push(treasure)
          haveStolen()
          return treasure
        end
     end
        return nil
  end
  
  def giveMeATreasure()
    aux= rand(@hiddenTreasures.length)
    return @hiddenTreasures.at(aux);
  end
  
  def canYouGiveMeATreasure()
    if(@hiddenTreasures.empty?)
      return false
    else
      return true
    end
  end
  
  def haveStolen()
    @canIsteal=false
  end

  def discardAllTreasures()
    auxVisibleTreasure =Array.new(@visibleTreasures)
    auxHiddenTreasure =Array.new(@hiddenTreasures)
        auxVisibleTreasure.each do |t|
            discardVisibleTreasure(t)
        end
        auxHiddenTreasure.each do |x|
            discardHiddenTreasure(x)
        end
  end
  
  def getOponentLevel(m)
    return m.combatLevel
  end
  
  def shouldConvert
    dice=Dice.instance
    number =dice.nextNumber
    if(number==1)
      return true
    else
      return false
    end
  end
  
  
  def to_s
   "\nNombre del Jugador : #{@name} \n Nivel: #{@level} \n Enemigo: #{@enemy.name} \n Tesoros Visibles: #{@visibleTreasures} \n Tesoros Ocultos: #{@hiddenTreasures} \n Mal Rollo Pendiente: #{@pendingBadConsequence} \n 
  \n "
  end
  
  private :bringToLife, :incrementLevels, :decrementLevels, :applyPrize, 
    :applyBadConsequence, :canMakeTreasureVisible, :howManyVisibleTreasures, :dielfNoTreasures, :giveMeATreasure,
     :haveStolen
  
  
  
  
end
end

File: napakalaki/src/Napakalaki/Player.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */

package Napakalaki;
import java.util.ArrayList;
import java.util.Random;
import GUI.Dice;


/**
 *
 * @author marcos0
 */
public class Player {
    static final int MAXLEVELS = 10;
    private String name;
    private int level;
    private boolean dead;
    private boolean canISteal;
    private BadConsequence pendingBadConsequence;
    protected Player enemy;
    private ArrayList<Treasure> visibleTreasure;
    private ArrayList<Treasure> hiddenTreasure;
    
    
    public Player(String _name){
        name=_name;
        level=1;
        dead=true;
        canISteal=true;
        pendingBadConsequence=null;
        enemy=null;
        visibleTreasure=new ArrayList();
        hiddenTreasure=new ArrayList();
    }
    
    public Player(Player p){
        name=p.name;
        level=p.level;
        dead=p.dead;
        canISteal=p.canISteal;
        pendingBadConsequence=p.pendingBadConsequence;
        enemy=p.enemy;
        visibleTreasure=p.visibleTreasure;
        hiddenTreasure=p.hiddenTreasure;
        
    }
  
    public String getName(){
        return name;
    }
    
    private void bringToLife(){
        dead = false;
    }
    
    protected int getCombatLevel(){ 
        int result=level;
        for (Treasure t: visibleTreasure){
            result+=t.getBonus();
        }
        return result;
    }
    
    public int getNivelCombate(){
        int aux=this.getCombatLevel();
        return aux;
    }
    
    private void incrementLevels(int l){
        if(level+l<=MAXLEVELS){
            level = level + l;
        }
        else{
            level=MAXLEVELS;
        }
    }
    
    protected void decrementLevels(int l){
        if(level-l>=0){
            level = level - l;
        }
        else {
            level=1;
            //dead=true;
        }
    }
    
    //EXAMEN
    public int quitarNivelesEnemigo(ArrayList<Treasure> selOcultos){
        int puntos=0;
        for(Treasure t : selOcultos){
            puntos+=t.getBonus();
        }
        if(puntos>5){
           puntos=4;
        }
        this.enemy.decrementLevels(puntos);
        return puntos;
    }
    //FIN EXAMEN
    
    private void setPendingBadConsequence(BadConsequence b){
        pendingBadConsequence = b; 
    }
    
    public BadConsequence getPendingBadConsequence(){
        return pendingBadConsequence;
    }
    private void applyPrize(Monster m){
        int nLevels=m.getLevelsGained();
        this.incrementLevels(nLevels);
        int nTreasures=m.getTreasuresGained();
        if(nTreasures>0){
            CardDealer dealer=CardDealer.getInstance();
            for(int i=0; i<nTreasures; i++){
                Treasure treasure = dealer.nextTreasure();
                hiddenTreasure.add(treasure);
            }
        }
    }
    
    private void applyBadConsequence(Monster m){
        BadConsequence bc = m.getBadConsequence();
        int nLevels = bc.getLevels();
        this.decrementLevels(nLevels);
        BadConsequence pendingBad = bc.adjustToFitTreasureList(visibleTreasure, hiddenTreasure);
        this.setPendingBadConsequence(pendingBad);
    }
    
    protected boolean canMakeTreasureVisible(Treasure t){
        int aux=0;
        boolean resultado=true;
        boolean tiene_dosmanos=false;
        for(Treasure v: visibleTreasure){
            if(v.getType()==TreasureKind.BOTHHANDS){
                tiene_dosmanos=true;
            }
            if(v.getType()==t.getType()){
                resultado=false;
            }
            if(v.getType()==TreasureKind.ONEHAND){
                aux++;
            }
        }
        if(t.getType()==TreasureKind.ONEHAND && aux<2){
            if(tiene_dosmanos==true){
                resultado=false;
            }
            else{
                resultado=true;
            }
        }
        if(aux>=1 && t.getType()==TreasureKind.BOTHHANDS){
            resultado=false;
        }
        return resultado;
    }
    
    private int howManyVisibleTreasure(TreasureKind tkind){
        int contador=0;
        for(Treasure t: visibleTreasure){
            if(t.getType()== tkind){
                contador++;
            }
        }
        return contador;
    }
    
    private void dieIfNoTreasures(){
        if(visibleTreasure.isEmpty() && hiddenTreasure.isEmpty()){
            dead=true;
        }
            
    }
    
    public boolean isDead(){
        return dead;     
    }
    
    public ArrayList<Treasure> getHiddenTreasures(){
        return hiddenTreasure;
    }
        
    public ArrayList<Treasure> getVisibleTreasures(){
        return visibleTreasure;
    }
    
    public CombatResult combat(Monster m){
        CombatResult aux;
        int myLevel = this.getCombatLevel();
        int monsterLevel = this.getOponentLevel(m);
        if(this.getLevels()>=MAXLEVELS){
            aux= CombatResult.WINGAME;
        }
        else if(myLevel>monsterLevel){
            this.applyPrize(m);
            aux = CombatResult.WIN;
            
        }
        else{
            this.applyBadConsequence(m);
            if(this.shouldConvert()){
                aux=CombatResult.LOSEANDCONVERT;
            }
            else{
                aux = CombatResult.LOSE;
            }
        }
        return aux;
    }
    
    public void makeTreasureVisible(Treasure t){
        boolean canl = this.canMakeTreasureVisible(t);
        if(canl){
            visibleTreasure.add(t);
            hiddenTreasure.remove(t);
        }
    }
    
    public void discardVisibleTreasure(Treasure t){
        visibleTreasure.remove(t);
        if(pendingBadConsequence!=null && !pendingBadConsequence.isEmpty()){
            pendingBadConsequence.substractVisibleTreasure(t);
        }
        this.dieIfNoTreasures();
    }
    
    public void discardHiddenTreasure(Treasure t){
        hiddenTreasure.remove(t);
        if(pendingBadConsequence!=null && !pendingBadConsequence.isEmpty()){
            pendingBadConsequence.substractHiddenTreasure(t);
        }
        this.dieIfNoTreasures();
    }
    
    public boolean validState(){
        boolean aux=true;
        if (pendingBadConsequence!=null){
            if(!pendingBadConsequence.isEmpty()){ 
                aux=false;
            }
        }
        if(hiddenTreasure.size()>4){
            aux=false;
        }
        return aux;
    }
    
    public void initTreasures(){
       CardDealer dealer = CardDealer.getInstance();
       Dice dice=Dice.getInstance();
       this.bringToLife();
       Treasure treasure = dealer.nextTreasure();
       hiddenTreasure.add(treasure);
       int number = dice.nextNumber("Tira para conseguir los tesoros","1->1 tesoro, mayor que 1->2 tesoros, 6->3 tesoros");
       if(number>1){
           treasure=dealer.nextTreasure();
           hiddenTreasure.add(treasure);
       }
       if(number==6){
           treasure = dealer.nextTreasure();
           hiddenTreasure.add(treasure);
       }
    }
    
    public int getLevels(){
        return level;
    }
    
    public Treasure stealTreasure(){
        boolean canl = this.canISteal();
        boolean canYou;
        Treasure treasure;
        if(canl){
            canYou = enemy.canYouGiveMeATreasure();
            if(canYou){
                treasure = enemy.giveMeATreasure();
                hiddenTreasure.add(treasure);
                this.haveStolen();
                return treasure;
            }
        }
        return null;
    }
    
    public void setEnemy(Player _enemy){
        enemy=_enemy;
    }
    
    protected Treasure giveMeATreasure(){
        Random generator=new Random();
        int aux=(generator.nextInt(hiddenTreasure.size()));
        return hiddenTreasure.get(aux);
    }
    
    public boolean canISteal(){
        return canISteal;
    }
    
    protected boolean canYouGiveMeATreasure(){
        if(hiddenTreasure.isEmpty()){
            return false;
        }
        else{
            return true;
        }
    }
    
    private void haveStolen(){
        canISteal=false;
    }
    
    public void discardAllTreasures(){
        ArrayList<Treasure> auxVisibleTreasure =new ArrayList(visibleTreasure);
        ArrayList<Treasure> auxHiddenTreasure =new ArrayList(hiddenTreasure);
        for(Treasure t: auxVisibleTreasure){
            this.discardVisibleTreasure(t);
        }
        for(Treasure x: auxHiddenTreasure){
            this.discardHiddenTreasure(x);
        }
    }
   
    protected int getOponentLevel(Monster m){
        return m.getCombatLevel();
    }
    
    protected boolean shouldConvert(){
        Dice dice=Dice.getInstance();
        int number =dice.nextNumber("Tira el dado para convertirte en sectario","Si sale 1 te conviertes!!");
        if(number==1){
            return true;
        }
        else{
            return false;
        }
    }
    
    public Player getEnemy(){
        return enemy;
    }
    
    @Override
    public String toString(){
        return "Nombre del jugador = " + (name) + "\n "
                + " Nivel del Jugador = " + Integer.toString(level) + "\n "
                + "Muerte  = " + dead +"\n "
                + "Mal rollo pendiente = " + pendingBadConsequence +"\n "
                + "Enemigo = " + enemy.getName() + "\n "
                + "Tesoros Visibles = " + visibleTreasure +"\n "
                + "Tesoros Ocultos = " + hiddenTreasure + " "
                + "\n";
          
    }
}
